### PR TITLE
Enable ecommerce on sandbox configuration.

### PR DESCRIPTION
**Pull request by agaylard**
This PR creates an ansible config to enable ecommerce features on sandbox.
*NB:* this PR has been re-based, and replaces PR-2.

**JIRA tickets**: Implements OC-1382

**Discussions**: This change

1. adds two new variables:
 - ```COMMON_ENABLE_ECOMMERCE```: enable ecommerce features.  This defaults to "false" to preserve the prior behaviour.
 - ```EDXAPP_ENABLE_OAUTH2_PROVIDER```: This allows the ```FEATURES``` list to be populated with the ```ENABLE_OAUTH2_PROVIDER```.  Previously, it was necessary to edit /edx/etc/edxapp/[cl]ms.env.yaml and add it manually.

2. modifies edx_sandbox.yml, adding the ecommerce and ecomworker roles, conditional on ```COMMON_ENABLE_ECOMMERCE```.

3. modifies roles/ecommerce/meta/main.yml, adding "ecommerce" to the list of nginx sites.

**Dependencies**: None

**Screenshots**: Not applicable.

**Sandbox URL**: dogwood.sandbox.opencraft.com

**Partner information**: Not applicable.

**Testing instructions**:

1. To build a sandbox with the ecommerce features enabled, use these changes together with a file of box-specific settings (not included here), like this:
```sudo /edx/bin/ansible-playbook -c local -i "localhost," edx_sandbox.yml --extra-vars=@edx_sandbox_ecommerce_dogwood.yml```

2. The edx_sandbox_ecommerce_dogwood file can be found attached to the ticket.  It is ready for testing on the dogwood.sandbox.opencraft.com computer.

3. Two manual steps are required after running this ansible command, before ecommerce can put sample transactions through PayPal.  They are outlined in the howto document at https://docs.google.com/document/d/1vkBewzwkizR5YndHM7AFHK5-HbFnFKVv3h2yBpDdYIw .

**Reviewers**
- [] Code: omarkhan
- [] Product: (TBD)
- [] Platform: (TBD)
- [] UX: Not applicable.
- [] TNL: Not applicable.
- [] a11y: Not applicable.
- [] doc/strings: Not applicable.

### CODE REVIEW

File: playbooks/roles/edxapp/defaults/main.yml [hunk 1]
```diff
@@ -132,6 +132,7 @@ EDXAPP_CAS_ATTRIBUTE_PACKAGE: ""
 EDXAPP_ENABLE_AUTO_AUTH: false
 # Settings for enabling and configuring third party authorization
 EDXAPP_ENABLE_THIRD_PARTY_AUTH: false
+EDXAPP_ENABLE_OAUTH2_PROVIDER: false
```
Comment (omarkhan) [inline playbooks/roles/edxapp/defaults/main.yml]: Why do we need to disable this here?

File: playbooks/edx_sandbox.yml [hunk 2]
```diff
@@ -37,6 +38,10 @@
       when: "'localhost' in EDXAPP_MONGO_HOSTS"
     - { role: 'edxapp', celery_worker: True }
     - edxapp
+    - role: ecommerce
+      when: COMMON_ENABLE_ECOMMERCE
+    - role: ecomworker
+      when: COMMON_ENABLE_ECOMMERCE
```
Comment (omarkhan) [inline playbooks/edx_sandbox.yml]: I think you included this twice by mistake

File: playbooks/roles/edxapp/defaults/main.yml (hunk 1 above)
Comment (agaylard) [inline playbooks/roles/edxapp/defaults/main.yml]: I disable this by default, because that is what the previous behaviour was.
For our purposes of enabling ecommerce, we override it in the supplementary settings file to enable it.

File: playbooks/edx_sandbox.yml (hunk 2 above)
Comment (agaylard) [inline playbooks/edx_sandbox.yml]: Er, I don't think so...?

These are two separate ansible roles.  They both need to be conditionally included.

Is there a less verbose way to do this?

Comment (omarkhan) [inline playbooks/edx_sandbox.yml]: Oops good point, I didn't look closely enough :smile: 

Maybe it would make sense to make the `ecommerce` role depend on the `ecomworker` role. What do you think?

File: playbooks/roles/edxapp/defaults/main.yml (hunk 1 above)
Comment (omarkhan) [inline playbooks/roles/edxapp/defaults/main.yml]: You mean this was recently removed and you are putting it back in? Can you add a comment to explain why this needs to be disabled by default in sandboxes?

File: playbooks/edx_sandbox.yml (hunk 2 above)
Comment (agaylard) [inline playbooks/edx_sandbox.yml]: I think it's best that these roles remain separate; that seems to be what the rest of the ansible config does with them.

Comment (omarkhan) [inline playbooks/edx_sandbox.yml]: Makes sense, I suppose you might want to have the worker running on a separate machine.

File: playbooks/roles/edxapp/defaults/main.yml (hunk 1 above)
Comment (agaylard) [inline playbooks/roles/edxapp/defaults/main.yml]: Sorry, my explanation wasn't clear.

This has never been included in the config.  Now that I'm adding it for the first time, my intention is that it's off for everybody but us.

When we want it, we can enable it specifically, by overriding it in our supplementary settings file.

Comment (omarkhan) [inline playbooks/roles/edxapp/defaults/main.yml]: Why do your changes here (adding the ecommerce roles) require that `EDXAPP_ENABLE_OAUTH2_PROVIDER ` be off by default?

Comment (agaylard) [inline playbooks/roles/edxapp/defaults/main.yml]: The reason it needs to be off by default, is that this file is used for **all** ansible configs, not just the sandbox, so we don't want to cause new behaviour for people not interested in the sandbox config.

When we want to run a playbook for sandbox + ecommerce (not a common setup, it seems),  then we will set it to "true" in edx_sandbox_ecommerce_dogwood.yml.

Hope that helps?

Comment (omarkhan) [inline playbooks/roles/edxapp/defaults/main.yml]: I understand that this is used in non-sandbox deployments. The default for this is already set in the edx-platform settings files. Unless this change has something to do with ecommerce specifically, I don't think it belongs in this PR.

Comment (agaylard) [inline playbooks/roles/edxapp/defaults/main.yml]: OK, I see your point — sorry for being so dense :)

Currently, this setting exists, and is set to True, in these files
- lms/envs/dev.py
- lms/envs/devstack.py
- lms/envs/test.py

However, lms/envs/common.py sets it to False.

To enable ecommerce to authenticate against LMS, it needs to be True.

So for those of us needing ecommerce and working in, say, devstack, nothing's needed.  

However for the sandbox config., we need to set it somewhere; I chose /edx/etc/edxapp/cms.env.yaml, which is written by roles/edxapp/templates/cms.env.yaml.j2, with variables populated from roles/edxapp/defaults/main.yml.

If you prefer, I can split this off to a separate PR.

**Comment by antoviaque** [comment]
@agaylard This is meant to be merged upstream, right? We usually open our PRs directly against upstream - this avoid having the reviews split up between two PRs, and makes it more simpler for everyone to discuss at the same place.

File: playbooks/roles/edxapp/defaults/main.yml (hunk 1 above)
Comment (omarkhan) [inline playbooks/roles/edxapp/defaults/main.yml]: Yes I think this should be in a separate PR.
